### ISSUE TITLE
Do we embrace Kernel#autoload ?

### ISSUE THREAD

**Issue by pitr-ch** [ruby-concurrency/concurrent-ruby#395]
Discussed here https://github.com/ruby-concurrency/concurrent-ruby/pull/386

**Comment by pitr-ch** [comment]
I would rather not use it. https://github.com/rails/rails/issues/17091#issuecomment-62419427 and https://github.com/headius/thread_safe/issues/15#issuecomment-36153742 mentions that it does not play well with different autoload mechanisms. Since we are only library we cannot control what else is pulled to the application. It may not work well together - needs further investigation.

**Comment by sschepens** [comment]
I don't see much benefit in using it, plus it implies overhead during runtime instead of on load and I almost never prefer that.

**Comment by thedarkone** [comment]
> plus it implies overhead during runtime

What do you mean?

**Comment by pitr-ch** [comment]
more random thoughts: if we fully migrate to autoload, users will only require 'concurrent', rest of the require points will be removed and its manually maintained require dependencies too. So for us it would be simpler.

**Comment by thedarkone** [comment]
Ad @pitr-ch's last comment:

While at #409 (in a now invisible @jdantonio's comment about some broken code) it just occurred to me, that he said he was running Rails and checking for warnings, but the code responsible for warnings was Rubinius code (that should only ever be required and evaled on rbx) and at the same time I was pretty certain that Rbx doesn't run Rails `master` tests right now (or at least it must be quite difficult to do). So it must have been that he was running MRI that was emitting warnings for Rbx specific `.rb` files, so it means `requires` are not 100% correct and while I'm pretty certain that this wouldn't ever occur with `autoload`.

It is unbelievably annoying to keep `require`s at absolute provable minimum (as in: "all `require` lines are necessary for library to load, removing even a single one will break things").

PS: `concurrent-ruby` is a much bigger lib than `thread_safe` ever was, so there might be some problems with `autoload` that never showed in `thread_safe`, but will pose a problem for `concurrent-ruby`.

**Comment by jdantonio** [comment]
@thedarkone You ate correct, I was getting that warning on MRI. The reason we were loading Rbx files on MRI was located in `utils.rb`, which I've updated in 525cd1801e77129866abe84706da0b1b07369e36.

As a general practice we treat `require`s as though this was Java, and endeavor to explicitly require all dependencies in every file. We then guard interpreter-specific files with `if` statements, rather than `autoload`. That's how this particular set of files slipped through. It's also the reason we have been getting circular reference warnings with the global configuration warning, but that's another issue.

**Comment by thedarkone** [comment]
Yes, I know that this was just a temporary post-merge issue. I was just saying that keeping a code base in such a state, that no `.rb` file is ever loaded unless absolutely necessary, is an unbelievably "hard problem" (as in nothing particularly bad happens, but on other hand no developer has the time to keep checking that *all the requires* at the top of the file are indeed needed).

**Comment by pitr-ch** [comment]
@thedarkone yeah it is hard, and I am afraid we have more than one bug in our requires currently since we do not test it. I like explicit requires more, but autoload just may be more practical. On the other hand we could explore some possibilities how to manage the requires: maybe parsing the constants out of a file, letting running Ruby process with required concurrent-ruby resolve the constant's full name and add requires based on that (Won't work in all cases but could be a good start).

**Comment by jdantonio** [comment]
I believe this problem is much simpler than that, and one that we've just about solved. The idea of reading constants out of a file, though novel, feels overly complex to me.

If we treat `require` as though this were Java/C#/etc. and every file `require`s every file it needs, then everything loads exactly as it's supposed to. `autoload` is not necessary--assuming no circular references. So we just need to get rid of circular references. Right now we only have two and they are both special cases.

The big problem is `Delay`. The original implementation was based on Clojure's [delay](http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/delay). Accessing the value of a Delay was a *blocking* operation. It was a synchronized operation that happened on the calling thread. The class was simple and had no internal dependencies. We then used Delay *within* the global configuration to lazy-load the internal thread pools. The problem occurred when a timeout was added to Delay during Actor development. To have a timeout another thread was needed, so the ability to inject an executor was added. For consistency with other classes, we allowed the global thread pools to be implicitly injected using the symbols `:io` and `:fast`. This is the exact moment when the circular dependency was introduced. The global configuration needs Delay to lazy-load, and Delay needs the global config to get its thread pools.

We can solve this entirely within Delay. Two approaches come to mind. The first is to revert to the *original* behavior of Delay, which aligns with Clojure, and make get rid of the timeout. The other is to remove the `:executor` option from the constructor and, instead, require an executor object to be injected at the time of the wait. Something similar to Clojure's [send-via](http://clojure.github.io/clojure/clojure.core-api.html#clojure.core/send-via) function. Basic Delay operation becomes synchronous/blocking again, but we add a `#wait_via` function which takes an executor as a argument. No convenience `:io` or `:fast`. In this case Delay won't need to `require` the global config or the executor classes any more. That `require` will be handled by the caller.

I will spike an implementation of this today and create a PR for review.

The other special case is `lib/concurrent/executor/executor.rb`, which I believe I've handled in PR #409. It's nothing more than a function to enable the aforementioned `:io` and `:fast` conveniences. It is a *pure function* with a `case` statement that returns the appropriate executor based on the given symbol. It's used by many of our high-level classes (like Delay) and depends on the global configuration. Because it is nothing more than a pure function on a module that is only ever called from within the constructor of other classes, it is perfectly safe to autoload. The simple rule of "always autoload Concurrent::Options" is all it will take to prevent problems.

**Comment by jdantonio** [comment]
I should also note that this is strictly a 1.0 discussion. The 2.0 release won't have any circular references. Since we are creating new APIs for many of the core high-level objects, we will define all those APIs in ways that do not create circular dependencies.

**Comment by pitr-ch** [comment]
> If we treat require as though this were Java/C#/etc. and every file requires every file it needs, then everything loads exactly as it's supposed to.

Yeah it's a simple and nice rule but it may be difficult to maintain in practice, it's a just a convention (error-prone) and we do not have a way to test it right now. I would like to go this way, but I do not see a clear and easy path.

Regarding Delay, I would like to keep the current API (similar to other IVar, Future, Promise), since the coming implementation of Delay in new futures is even more integrated. I would suggest to create simplistic internal implementation just to solve the circular dependency.

**Comment by jeremy** [comment]
On another tack, autoload can help with boot time and heap size. +1 to both explicit requires (every file requires every file it needs) for internal deps as well as embracing autoload for toplevel constants to limit footprint.

Example [profiling require time](https://github.com/rails/rails/blob/master/tools/profile):
```bash
fwip 2.2.3 ~/rails master= tools/profile activesupport/lib/active_support/values/time_zone.rb
             /Users/jeremy/rails/activesupport/lib/active_support/values/time_zone.rb
              tzinfo
               tzinfo/ruby_core_support
                date
     4.1 ms      date_core
     6.2 ms     date
     8.9 ms    tzinfo/ruby_core_support
     1.5 ms    tzinfo/offset_rationals
               tzinfo/time_or_datetime
     4.8 ms     time
     8.6 ms    tzinfo/time_or_datetime
     0.9 ms    tzinfo/timezone_definition
     1.1 ms    tzinfo/timezone_offset
     1.5 ms    tzinfo/timezone_transition
     1.1 ms    tzinfo/timezone_transition_definition
     1.0 ms    tzinfo/timezone_index_definition
     0.9 ms    tzinfo/timezone_info
     0.9 ms    tzinfo/data_timezone_info
     0.9 ms    tzinfo/linked_timezone_info
     2.1 ms    tzinfo/transition_data_timezone_info
     2.1 ms    tzinfo/zoneinfo_timezone_info
     1.9 ms    tzinfo/data_source
     1.2 ms    tzinfo/ruby_data_source
     2.7 ms    tzinfo/zoneinfo_data_source
     1.3 ms    tzinfo/timezone_period
     4.6 ms    tzinfo/timezone
     0.8 ms    tzinfo/info_timezone
     0.9 ms    tzinfo/data_timezone
     0.9 ms    tzinfo/linked_timezone
     1.0 ms    tzinfo/timezone_proxy
     0.9 ms    tzinfo/country_index_definition
     0.9 ms    tzinfo/country_info
     1.0 ms    tzinfo/ruby_country_info
     0.9 ms    tzinfo/zoneinfo_country_info
     1.5 ms    tzinfo/country
     1.1 ms    tzinfo/country_timezone
    55.3 ms   tzinfo
              concurrent
     0.2 ms    concurrent/version
     0.2 ms    concurrent/constants
     0.3 ms    concurrent/errors
               concurrent/configuration
                concurrent/delay
                 concurrent/concern/obligation
     1.3 ms       timeout
                  concurrent/atomic/event
                   concurrent/synchronization
     0.3 ms         concurrent/utility/engine
     0.3 ms         concurrent/synchronization/abstract_object
                    concurrent/utility/native_extension_loader
                     concurrent/extension
                      concurrent/2.2/extension
   126.0 ms           concurrent/utility/native_extension_loader
     0.5 ms           concurrent/synchronization/mri_object
     0.3 ms           concurrent/synchronization/jruby_object
     0.3 ms           concurrent/synchronization/rbx_object
     0.6 ms           concurrent/synchronization/object
     0.3 ms           concurrent/synchronization/abstract_lockable_object
     0.4 ms           concurrent/synchronization/mri_lockable_object
     0.2 ms           concurrent/synchronization/jruby_lockable_object
     0.4 ms           concurrent/synchronization/rbx_lockable_object
     0.3 ms           concurrent/synchronization/lockable_object
     0.4 ms           concurrent/synchronization/condition
     0.3 ms           concurrent/synchronization/lock
   130.8 ms          concurrent/synchronization
   131.7 ms         concurrent/atomic/event
     0.4 ms         concurrent/concern/dereferenceable
   134.6 ms        concurrent/concern/obligation
                   concurrent/executor/immediate_executor
                    concurrent/executor/abstract_executor_service
                     concurrent/executor/executor_service
                      concurrent/concern/logging
     2.9 ms            logger
     3.1 ms           concurrent/concern/logging
     3.4 ms          concurrent/executor/executor_service
     1.1 ms          concurrent/utility/at_exit
     5.2 ms         concurrent/executor/abstract_executor_service
     0.2 ms         concurrent/executor/serial_executor_service
     5.8 ms        concurrent/executor/immediate_executor
   141.6 ms       concurrent/delay
                  concurrent/atomic/atomic_reference
     0.2 ms        concurrent/atomic_reference/concurrent_update_error
                   concurrent/atomic_reference/mutex_atomic
     0.3 ms         concurrent/atomic_reference/direct_update
     0.2 ms         concurrent/atomic_reference/numeric_cas_wrapper
     0.9 ms        concurrent/atomic_reference/mutex_atomic
     0.2 ms        concurrent/atomic_reference/ruby
     1.7 ms       concurrent/atomic/atomic_reference
     1.6 ms       concurrent/utility/processor_counter
   146.3 ms      concurrent/configuration
                 concurrent/atomics
                  concurrent/atomic/atomic_boolean
     0.4 ms        concurrent/atomic/mutex_atomic_boolean
     0.6 ms       concurrent/atomic/atomic_boolean
                  concurrent/atomic/atomic_fixnum
     0.5 ms        concurrent/atomic/mutex_atomic_fixnum
     0.8 ms       concurrent/atomic/atomic_fixnum
     0.5 ms       concurrent/atomic/cyclic_barrier
                  concurrent/atomic/count_down_latch
     0.3 ms        concurrent/atomic/mutex_count_down_latch
     0.3 ms        concurrent/atomic/java_count_down_latch
     0.9 ms       concurrent/atomic/count_down_latch
     1.3 ms       concurrent/atomic/read_write_lock
                  concurrent/atomic/semaphore
     0.5 ms        concurrent/atomic/mutex_semaphore
     0.8 ms       concurrent/atomic/semaphore
                  concurrent/atomic/thread_local_var
                   concurrent/atomic/ruby_thread_local_var
     0.3 ms         concurrent/atomic/abstract_thread_local_var
     1.5 ms        concurrent/atomic/ruby_thread_local_var
     0.3 ms        concurrent/atomic/java_thread_local_var
     2.1 ms       concurrent/atomic/thread_local_var
     7.4 ms      concurrent/atomics
                 concurrent/executors
                  concurrent/executor/cached_thread_pool
                   concurrent/executor/thread_pool_executor
                    concurrent/executor/ruby_thread_pool_executor
     0.4 ms          concurrent/executor/ruby_executor_service
     0.4 ms          concurrent/utility/monotonic_time
     2.4 ms         concurrent/executor/ruby_thread_pool_executor
     2.7 ms        concurrent/executor/thread_pool_executor
     3.0 ms       concurrent/executor/cached_thread_pool
     0.4 ms       concurrent/executor/fixed_thread_pool
                  concurrent/executor/indirect_immediate_executor
     0.5 ms        concurrent/executor/simple_executor_service
     0.8 ms       concurrent/executor/indirect_immediate_executor
     0.5 ms       concurrent/executor/java_executor_service
     0.3 ms       concurrent/executor/java_single_thread_executor
     0.5 ms       concurrent/executor/java_thread_pool_executor
     0.3 ms       concurrent/executor/ruby_single_thread_executor
     0.3 ms       concurrent/executor/safe_task_executor
     0.6 ms       concurrent/executor/serialized_execution
     0.9 ms       concurrent/executor/serialized_execution_delegator
     0.3 ms       concurrent/executor/single_thread_executor
                  concurrent/executor/timer_set
                   concurrent/scheduled_task
                    concurrent/ivar
     0.5 ms          concurrent/collection/copy_on_write_observer_set
                     concurrent/concern/observable
     0.5 ms           concurrent/collection/copy_on_notify_observer_set
     0.8 ms          concurrent/concern/observable
     2.1 ms         concurrent/ivar
     2.9 ms        concurrent/scheduled_task
                   concurrent/collection/non_concurrent_priority_queue
     0.4 ms         concurrent/collection/java_non_concurrent_priority_queue
     0.9 ms         concurrent/collection/ruby_non_concurrent_priority_queue
     1.6 ms        concurrent/collection/non_concurrent_priority_queue
     5.3 ms       concurrent/executor/timer_set
    13.6 ms      concurrent/executors
     2.3 ms      concurrent/agent
     1.1 ms      concurrent/atom
                 concurrent/array
     0.4 ms       concurrent/thread_safe/util
     0.9 ms      concurrent/array
     0.5 ms      concurrent/hash
                 concurrent/map
                  concurrent/collection/map/mri_map_backend
     0.8 ms        concurrent/collection/map/non_concurrent_map_backend
     3.3 ms       concurrent/collection/map/mri_map_backend
     5.6 ms      concurrent/map
     0.8 ms      concurrent/tuple
     0.8 ms      concurrent/async
                 concurrent/dataflow
     1.2 ms       concurrent/future
     1.7 ms      concurrent/dataflow
                 concurrent/exchanger
     0.5 ms       concurrent/maybe
     1.2 ms      concurrent/exchanger
                 concurrent/immutable_struct
     0.7 ms       concurrent/synchronization/abstract_struct
     1.2 ms      concurrent/immutable_struct
     0.7 ms      concurrent/mutable_struct
     0.6 ms      concurrent/mvar
     1.9 ms      concurrent/promise
     0.7 ms      concurrent/settable_struct
     0.8 ms      concurrent/timer_task
     1.4 ms      concurrent/tvar
     1.4 ms      concurrent/thread_safe/synchronized_delegator
     0.3 ms      concurrent/options
   197.5 ms     concurrent
     0.3 ms     active_support/core_ext/object/blank
     1.5 ms     active_support/core_ext/object/try
   258.0 ms    /Users/jeremy/rails/activesupport/lib/active_support/values/time_zone.rb
   258.0 ms  9200 KB RSS
```

**Comment by jdantonio** [comment]
@jeremy For the abstractions Rails is using you should be fine requiring just the things that are being used. `require 'concurrent/map'` and `require 'concurrent/atomic/count_down_latch'`. Neither of those use the global configuration, which is the one problem area. I'm pretty sure that `require 'concurrent/executor/thread_pool_executor'` will work in AsyncJob, too. Since that is only autoloaded by Rails when used that should be less of an issue.

I'll do some experimenting tonight and submit a PR to Rails if that works.

**Comment by jeremy** [comment]
@jdantonio Nice, thank you. I used the Rails case since it's easy to profile and a good example of how omitting requires internally can imply that only toplevel requires are supported. The specific case where I noticed this was wanting to `require 'concurrent/synchronization/lock'` directly, but it fails since it depends on LockableObject superclass and, looking at the file, it's hard for a user of the library to know which paths may be, much less should be, "safely" required: `concurrent/synchronization`, `concurrent`, or some other "blessed" entry points?

Knowing this, a user of the library can properly pull in the feature they need by either
1. Requiring the appropriate toplevel file, e.g. `concurrent/synchronization` or `concurrent` itself, then relying on autoload to pull in sub-features from there, or
2. Deep-requiring each specific feature, e.g. `concurrent/synchronization/lock`, and expecting it to require all its deps without hitting circular dep headaches.

In the first case, we'd want `concurrent` and other omnibus requires to use autoload to limit the lib footprint to only what's used. In the second case, we'd want to more clearly discourage use of toplevel omnibus requires like `concurrent` and steer users toward fine-grained specific requires.

Rails' components feel both pressures: fast boot and a simple usage model are desirable for development with quick startup, but thread-safe requires and CoW-friendliness are desirable for deployed services and preforking daemons. Hence the `eager_autoload` scheme to accommodate both: https://github.com/rails/rails/blob/7f18ea14c893cb5c9f04d4fda9661126758332b5/activesupport/lib/active_support/dependencies/autoload.rb

**Comment by thedarkone** [comment]
Maybe `concurrent-ruby` should depend on `rails` and use `ActiveSupport::Autoload`? :stuck_out_tongue_closed_eyes:

Sorry guys, couldn't resist :innocent:.

**Comment by jdantonio** [comment]
@jeremy I've created a branch on my personal Rails for in which I've [changed the require statements](https://github.com/jdantonio/rails/commit/218aa0f56f16ca7f17451c3978e115619022591d) but I can't get my output from `tools/profile` to display the same output as you. I've tried on both `master` and my working branch but I don't see concurrent-ruby in the output. What's the trick? What am I missing?

Here is what I get when I run the tool:

```
[22:02:22 jerry ~/Projects/rails (cr-load-time)]$ tools/profile activesupport/lib/active_support/values/time_zone.rb 
             /home/jerry/Projects/rails/activesupport/lib/active_support/values/time_zone.rb
              tzinfo
               tzinfo/ruby_core_support
                date
     0.4 ms      date_core
     0.9 ms     date
     1.5 ms    tzinfo/ruby_core_support
     0.5 ms    tzinfo/offset_rationals
               tzinfo/time_or_datetime
     2.3 ms     time
     3.2 ms    tzinfo/time_or_datetime
     0.2 ms    tzinfo/timezone_definition
     0.3 ms    tzinfo/timezone_offset
     0.4 ms    tzinfo/timezone_transition
     0.3 ms    tzinfo/timezone_transition_definition
     0.3 ms    tzinfo/timezone_index_definition
     0.2 ms    tzinfo/timezone_info
     0.2 ms    tzinfo/data_timezone_info
     0.2 ms    tzinfo/linked_timezone_info
     0.8 ms    tzinfo/transition_data_timezone_info
     1.0 ms    tzinfo/zoneinfo_timezone_info
     0.5 ms    tzinfo/data_source
     0.5 ms    tzinfo/ruby_data_source
     1.3 ms    tzinfo/zoneinfo_data_source
     0.6 ms    tzinfo/timezone_period
               tzinfo/timezone
                thread_safe
     0.2 ms      thread_safe/version
                 thread_safe/synchronized_delegator
     1.3 ms       delegate
     1.6 ms      thread_safe/synchronized_delegator
     2.1 ms     thread_safe
     4.9 ms    tzinfo/timezone
     0.2 ms    tzinfo/info_timezone
     0.2 ms    tzinfo/data_timezone
     0.2 ms    tzinfo/linked_timezone
     0.2 ms    tzinfo/timezone_proxy
     0.2 ms    tzinfo/country_index_definition
     0.2 ms    tzinfo/country_info
     0.3 ms    tzinfo/ruby_country_info
     0.2 ms    tzinfo/zoneinfo_country_info
     0.5 ms    tzinfo/country
     0.4 ms    tzinfo/country_timezone
    21.8 ms   tzinfo
     0.3 ms   active_support/core_ext/object/blank
     0.3 ms   active_support/core_ext/object/try
    24.0 ms  /home/jerry/Projects/rails/activesupport/lib/active_support/values/time_zone.rb
    24.0 ms  2328 KB RSS
```

**Comment by jeremy** [comment]
@jdantonio Check that the file has `require 'concurrent'` – added in your commit 😁 https://github.com/rails/rails/commit/56ac6e4768adb1f7055474d40a9e921380559c43

**Comment by jdantonio** [comment]
@jeremy It appears that the profiler may not output anything for files that load very fast. The [branch I'm working on](https://github.com/jdantonio/rails/commit/218aa0f56f16ca7f17451c3978e115619022591d) replaces `require 'concurrent'` with `require 'concurrent/map'`. Map does a lot of autoloading. When I do `require 'concurrent'` I get output identical to yours. I experimented with several other requires and some *very* light files, like `concurrent/errors` don't show up. Other files do.

Since all the Rails tests pass I'm going to create a PR to Rails. Thanks for the suggestion!

**Comment by pitr-ch** [comment]
@jeremy I understand your second point, but I think it's not a maintainable option for us, as discussed further up. It would be too error-prone and we do not have a way to test it. I think best option is to allow only `require 'concurrent'` and streamline the lib with autoload for top-level abstractions.

**Comment by jeremy** [comment]
Agree 👍

**Comment by pitr-ch** [comment]
On the other hand `autoload` presents a problem for factory methods like `Concurrent.future { task }` they would be added only after Concurrent::Future is hit somewhere.

**Comment by pitr-ch** [comment]
And that can be fixed with https://gist.github.com/pitr-ch/d31cf9787e7a4a392d6c, so using autoload is still possible,

**Comment by jdantonio** [comment]
I think we should abandon `autoload` for now and focus on making the gem fully support limited, targeted `require` statements (for example, `require 'concurrent/future'`. I've expanded on this in #520. Since this is a 2.0 discussion we can discuss it further on that issue.

**Comment by pitr-ch** [comment]
@jdantonio this issue is assigned in 2.0.0 milestone, for clarity I'll bring the discussion about autoload back.

Jerry said in #520:
> * The more I think about `autoload` the more I dislike it.
>  - Originally we allowed users to require specific subsets of the gem (for example, `require 'concurrent/future`). We only began using `autoload` when we merged in other gems that already used it.
>  - We moved away from subset requires only because of circular references we introduced with the global config.
>  - Since them we've cleaned things up and subset requires are possible again. 
>  - In 2.x we're going to simplify the global config and fix the circular requires so the problem will solve itself.
>  - Subset requires are explicit whereas `autoload` is implicit. The former fits better with our values.
>  - If we build 2.0 so that it fully support subset requires, we can easily add `autoload` later.

I am sorry for reopening the issue but I believe there are issues left to be discussed.

Explicit requires are good, in fact I like it more than autoload (possibly accompanied by autoload_methods). But before we go that path we should have a solution for following issues:
- how to prevent users from requiring bad files, giving them a friendly error
- how to test and ensure that all our files designed to be required have all the dependencies they need required themselves

Autoload makes sense just because it does not have these issues since it's implicit.

**Comment by jdantonio** [comment]
> how to prevent users from requiring bad files, giving them a friendly error

I don't believe we need to. Ruby already provides `LoadError` specifically for this purpose. Let's take ActiveSupport as an example:

```
[19:11:37 jerry ~/Projects/RailsBridge/suggestotron (master)]$ bundle console
2.2.3 :001 > require 'active_support/core_ext'
 => true 
2.2.3 :002 > require 'active_support/foo'
LoadError: cannot load such file -- active_support/foo
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/activesupport-4.1.5/lib/active_support/dependencies.rb:247:in `require'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/activesupport-4.1.5/lib/active_support/dependencies.rb:247:in `block in require'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/activesupport-4.1.5/lib/active_support/dependencies.rb:232:in `load_dependency'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/activesupport-4.1.5/lib/active_support/dependencies.rb:247:in `require'
	from (irb):2
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/cli/console.rb:14:in `run'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/cli.rb:319:in `console'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/cli.rb:10:in `start'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/exe/bundle:19:in `block in <top (required)>'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/lib/bundler/friendly_errors.rb:7:in `with_friendly_errors'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/gems/bundler-1.11.2/exe/bundle:17:in `<top (required)>'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/bin/bundle:22:in `load'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/bin/bundle:22:in `<main>'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/bin/ruby_executable_hooks:15:in `eval'
	from /home/jerry/.rvm/gems/ruby-2.2.3@railsbridge/bin/ruby_executable_hooks:15:in `<main>'2.2.3 :003 > 
```

ActiveSupport is designed to allow users to include specific subsets of the gem. When users try to require something that doesn't exist it simply let's Ruby raise its normal exception. If that's good enough for ActiveSupport it's good enough for me.

> how to test and ensure that all our files designed to be required have all the dependencies they need required themselves

Automated tests. We've talked a few times in the past about automating tests beyond just our simple unit tests. The tests I wrote to validate the packaged gems, our performance benchmarks, tests for memory leaks, etc. For 2.0 we have many opportunities for expand our automated testing and we should definitely explore them.

**Comment by jdantonio** [comment]
We will not embrace `autoload`. 2.0 will use explicit requires and will allow users to require only the parts they need. We will not use `autoload` in 2.0.
